### PR TITLE
EVG-13889 fetch by instance type

### PR DESCRIPTION
**Pull request by ybrill**
Jira: https://jira.mongodb.org/browse/EVG-13889

When you ask AWS for all the instance types supported in an AZ they seem to leave some out. For example, if you run this from the command line
```
aws ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=availability-zone,Values=us-east-1a
```
the list returned doesn't include a1.2xlarge.

Getting all the AZs that support a particular instance type works, though
```
aws ec2 describe-instance-type-offerings --location-type availability-zone --filters Name=instance-type,Values=a1.2xlarge
```

The downside to doing it this way is that we need to do one API call for each instance type as opposed to one for each AZ. This might not make such a difference, though, since distros use the same instance type.

### CODE REVIEW

File: cloud/ec2_fleet.go [hunk 1]
```diff
@@ -20,61 +20,62 @@ import (
...
-			if offering != nil && offering.InstanceType != nil {
-				c.instanceTypeToSubnets[*offering.InstanceType] = append(c.instanceTypeToSubnets[*offering.InstanceType], subnet)
-			}
+	if output == nil {
+		return nil, errors.Errorf("DescribeInstanceTypeOfferings returned nil output for AZ '%s' in '%s'", instanceRegion.instanceType, instanceRegion.region)
+	}
+	supportingAZs := make([]string, 0, len(output.InstanceTypeOfferings))
+	for _, offering := range output.InstanceTypeOfferings {
+		if offering != nil && offering.Location != nil {
+			supportingAZs = append(supportingAZs, *offering.Location)
```
Comment (ablack12) [inline cloud/ec2_fleet.go:74]: Don't we only want to include availability zones that are in the requested region? It doesn't look like we do anything with instanceRegion.region

**Review by ablack12 — CHANGES_REQUESTED** [review]

File: cloud/ec2_fleet.go [hunk 2]
```diff
@@ -20,61 +20,62 @@ import (
 )
 
 type instanceTypeSubnetCache struct {
```
[the remark this answers is not in the document]
Comment (ybrill) [reply cloud/ec2_fleet.go]: that's a great idea!
done

File: cloud/ec2_fleet.go (hunk 1 above)
Comment (ybrill) [reply cloud/ec2_fleet.go:72]: The region isn't used explicitly in the call, but the response will only include the AZs in the client's region.

For example, if we call it with a client configured with the us-east-1 region and get the AZs there I want that the next time we call it with a client configured for us-east-2 we should know that we haven't called it for this region yet and make the API call.

Comment (ablack12) [reply cloud/ec2_fleet.go:72]: Oh I see what you mean. Can you put a comment to that effect where DescribeInstanceTypeOfferings is called then, so it's clear that region is implicitly being considered?

**Review by ablack12 — APPROVED** [review]
LGTM with comment request

**Pull request merged**